File: week_2/day_08_stack/next_greater_follow_up_questions/nextGreaterCircular.js
function nextGreaterCircular(nums) {
    const n = nums.length;
    const res = new Array(n).fill(-1);
    const stack = [];

    for (let i = 0; i < 2 * n; i++) {
        const curr = nums[i % n];
        while (stack.length && nums[stack[stack.length - 1]] < curr) {
            const idx = stack.pop();
            res[idx] = curr;
        }
        if (i < n) stack.push(i);
    }
    return res;
}

// INPUT
console.log(nextGreaterCircular([1, 2, 1])); // [2, -1, 2]

File: week_2/day_08_stack/next_greater_follow_up_questions/next_greater_circular.py
def nextGreaterCircular(nums):
    n = len(nums)
    res = [-1] * n
    stack = []

    for i in range(2 * n):
        curr = nums[i % n]
        while stack and nums[stack[-1]] < curr:
            idx = stack.pop()
            res[idx] = curr
        if i < n:
            stack.append(i)

    return res

# INPUT
nums = [1, 2, 1]
print(nextGreaterCircular(nums))  # Output: [2, -1, 2]
